cpp/include/raft/comms/comms.hpp

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <memory>
+
+namespace raft {
+namespace comms {
+
+typedef unsigned int request_t;
+enum datatype_t { CHAR, UINT8, INT, UINT, INT64, UINT64, FLOAT, DOUBLE };
+enum op_t { SUM, PROD, MIN, MAX };
+
+/**
+ * The resulting status of distributed stream synchronization
+ */
+enum status_t {
+  commStatusSuccess,  // Synchronization successful
+  commStatusError,    // An error occured querying sync status
+  commStatusAbort     // A failure occurred in sync, queued operations aborted
+};
+
+class comms_iface {
+ public:
+  virtual ~comms_iface();
+
+  virtual int getSize() const = 0;
+  virtual int getRank() const = 0;
+
+  virtual std::unique_ptr<comms_iface> commSplit(int color, int key) const = 0;
+
+  virtual void barrier() const = 0;
+
+  virtual status_t syncStream(cudaStream_t stream) const = 0;
+
+  virtual void isend(const void* buf, int size, int dest, int tag,
```

Comment:
This follows the MPI API, which uses a 32-bit integer for this field.

Reply:
MPI uses int, but it is more like MPI's technical debt, and in theory, we'd better not follow their mistake.

In practice, if we rely on MPI to implement isend and if the size exceeds std::numerical_limit<int>::max(), we need to split a single isend call to multiple MPI non-blocking calls and we need to manage multiple request_t objects, and this will make our public API messier.

I suggest this. Use size_t size and if the backend is UCX or NCCL, this should just work. If the backend is MPI, if the size exceeds std::nemerical_limit<int>::max(), this call throws an exception; in this case, users will be responsible for splitting one large isend call to multiple isend calls.

Reply:
Just to be clear- I really don't mind this change at all, and it looks like both UCX and NCCL support `size_t` as well.